stgit: 1.5 -> 2.3.0

skyfmmf opened the pull request:
###### Description of changes

Version 2 of stgit was rewritten in Rust but in general offers the same feature set. We therefore do not require a Python build environment anymore but a Rust one. However, we overwrite the default build, check and install phases to use the Makefiles provided by stgit. In addition, the following changes are necessary:

- stgit now requires libcurl
- perl is required during build (to build docs correctly) and during check (to run the extensive test suite)

Changelog for v2.0.0: https://github.com/stacked-git/stgit/releases/tag/v2.0.0
Changelog for v2.3.0: https://github.com/stacked-git/stgit/releases/tag/v2.3.0
List of changelogs v2.0.0...v2.3.0: https://stacked-git.github.io/changelog/

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable: (checkPhase runs the extensive test suite and I tested a few commands locally)
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage) (nothing depends on stgit)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

skyfmmf committed:
stgit: 1.5 -> 2.3.0

Version 2 of stgit was rewritten in Rust but in general offers the same feature
set [1]. We therefore do not require a Python build environment anymore but a
Rust one. However, we overwrite the default build, check and install phases to
use the Makefiles provided by stgit. In addition, the following changes are
made:

- stgit now requires libcurl
- perl is required during build (to build docs correctly) and during check (to
  run the extensive test suite)
- the tests require sysctl and iconv on darwin; add them to the check inputs

The added patch fixes failing test due to the recent git version in nixpkgs.
The fix is also upstream and will most likely be released with 2.3.1, i.e., the
patch can be removed then.

[1]: https://github.com/stacked-git/stgit/releases/tag/v2.0.0

jshholland reviewed (approved):
LGTM; tested on x86_64-linux and aarch64-linux.

skyfmmf commented:
The PR needed a rebase as d994dabacef ("treewide: remove git-and-tools directory") was applied in the meantime, which moved the stgit package. The contents of the PR are still the same.

SuperSandro2000 commented on pkgs/applications/version-management/stgit/default.nix:
```diff
@@ -1,32 +1,37 @@
-  nativeBuildInputs = [ installShellFiles asciidoc xmlto docbook_xsl docbook_xml_dtd_45 python3Packages.setuptools ];
+  nativeBuildInputs = [ pkg-config installShellFiles asciidoc xmlto docbook_xsl docbook_xml_dtd_45 perl ];
+  buildInputs = [ libgit2 curl bzip2 ];
+  propagatedBuildInputs = [ git ];
```
Please use makeWrapper instead.

skyfmmf replied:
You mean, I should keep git in `buildInputs` and run makeWrapper/wrapProgram with `--prefix PATH : ${lib.makeBinPath [ git ]}` in `installPhase` instead?

skyfmmf replied:
Updated to use makeProgram

SuperSandro2000 commented on pkgs/applications/version-management/stgit/default.nix:
```diff
@@ -40,24 +45,20 @@ python3Packages.buildPythonApplication rec {
 
-  checkTarget = "test";
-  checkFlags = [ "PERL_PATH=${perl}/bin/perl" ];
+  checkPhase = [ "make" "test" ] ++ makeFlags;
```
This should be checkTarget

skyfmmf replied:
I'm not sure how that would work. This now uses `rustPlatform.buildRustPackage` which ignores checkTargets. Instead, it just runs the Rust test then, which are very different than the integration tests behind `make test`.

For a Rust based program that use a Makefiles, the nixpkgs docs write the following: 

> Some packages may use custom scripts for building/installing, e.g. with a Makefile. In these cases, it’s recommended to override the buildPhase/installPhase/checkPhase. 

(from https://nixos.org/manual/nixpkgs/stable/#compiling-rust-applications-with-cargo, 17.31.1.6. Custom build/install-procedures)

Using checkTarget that automatically uses the makeFlags sounds cleaner, I'm just now sure how I would use that with Rust.

skyfmmf replied:
As this is the recommended way to package Rust programs with custom build processes, I'll stick with checkPhase. But I updated it to be a multiline string as you noted.

SuperSandro2000 commented on pkgs/applications/version-management/stgit/default.nix:
```diff
@@ -40,24 +45,20 @@ python3Packages.buildPythonApplication rec {
+  checkPhase = [ "make" "test" ] ++ makeFlags;
 
-  installTargets = [ "install" "install-doc" "install-html" ];
+  installPhase = [ "make" "install" "install-man" "install-html" ] ++ makeFlags;
```
This should be kept installTargets. installPhase and checkPhase are multiline strings, not lists.

skyfmmf replied:
See my comment for checkPhase/checkTargets

skyfmmf replied:
As this is the recommended way to package Rust programs with custom build processes, I'll stick with installPhase. But I updated it to be a multiline string as you noted.

AndersonTorres commented:
Still needs rebase.

Janik-Haag reviewed (commented):
@skyfmmf can you please rebase so we can get this merged?

skyfmmf commented:
@Janik-Haag Sorry for loosing track of this for a while. 

I rebased, updated directly to the most recently released version and addressed the review comments.

Janik-Haag commented:
Result of `nixpkgs-review pr 205220` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>stgit</li>
  </ul>
</details>

Janik-Haag commented:
Darwin tests are failing, if you know how to fix them fix them, if not ping the darwin-maintainers team or make darwin as broken since it's supposed to be working.

skyfmmf commented:
@Janik-Haag I fixed the darwin tests. Apparently they need sysctl (to get the number of cores) and iconv to pass.

SuperSandro2000 commented on pkgs/applications/version-management/stgit/default.nix:
```diff
@@ -1,32 +1,48 @@
+  nativeCheckInputs = [
+    git perl
+  ] ++ lib.optionals stdenv.isDarwin [
+    darwin.system_cmds darwin.libiconv
```
this should be just libiconv

SuperSandro2000 commented on pkgs/applications/version-management/stgit/default.nix:
```diff
@@ -40,25 +56,28 @@ python3Packages.buildPythonApplication rec {
   ];
 
-  buildFlags = [ "all" "doc" ];
+  buildPhase = ''
+    make all ${makeFlags}
+  '';
+
+  checkPhase = ''
+    make test ${makeFlags}
+  '';
```
We don't need to manually define this, it should just work with makeFlags. Also runHook is missing

skyfmmf replied:
I just tried, it doesn't. But it turns out, most of the flags are not necessary anymore, I'll clean them up and add the runHooks with the next release of stgit
